File: blind75/linkedList/21-mergeTwoSortedLists.js
// 21. Merge Two Sorted Lists
// You are given the heads of two sorted linked lists list1 and list2.
// Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.
// Return the head of the merged linked list.


/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    let dummy = new ListNode();
    let tail = dummy;
    let head1 = list1;
    let head2 = list2;

    while(head1 !== null && head2 !== null) {
        if (head1.val < head2.val) {
            tail.next = head1;
            head1 = head1.next;
        } else {
            tail.next = head2;
            head2 = head2.next;
        }
        tail = tail.next;
    }

    if (head1 === null) 
        tail.next = head2;
    else if (head2 === null)
        tail.next = head1;
    return dummy.next;
};

// 这道题思路是 首先得有一个dummyNode 就是一个空的node
// 避免考虑那些特殊情况
// 然后再来遍历两个list，谁大选谁
// 最后当一个list到最后了只需要再拼接剩下的


let list1 = [1,2,4], list2 = [1,3,4];
let list1a = [], list2a = [];
let list1b = [], list2b = [0];

console.log(mergeTwoLists(list1, list2));
console.log(mergeTwoLists(list1a, list2a));
console.log(mergeTwoLists(list1b, list2b));

File: specified-learning专项训练/linkedList/21-mergeTwoSortedLists.py
# 21. Merge Two Sorted Lists
# You are given the heads of two sorted linked lists list1 and list2.
# Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.
# Return the head of the merged linked list.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

from typing import Optional
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        # tail 标记最末尾的值
        tail = dummy
        while list1 != None and list2 != None:
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
        # 如果有一个list到了结尾，可以直接用tail连着剩下的那个list，因为他本来就是有序的
        if list1 != None:
            tail.next = list1
        elif list2 != None:
            tail.next = list2
        return dummy.next

    # 我自己的做法
    # def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    #     dummy = ListNode()
    #     aNode = dummy
    #     while list1 != None and list2 != None:
    #         if list1.val <= list2.val:
    #             aNode.next = list1
    #             aNode = aNode.next
    #             list1 = list1.next
    #         else:
    #             aNode.next = list2
    #             aNode = aNode.next
    #             list2 = list2.next
    #     while list2 != None:
    #         aNode.next = list2
    #         aNode = aNode.next
    #         list2 = list2.next
    #     while list1 != None:
    #         aNode.next = list1
    #         aNode = aNode.next
    #         list1 = list1.next
            
    #     return dummy.next